data/src/buffer_ref.rs

```diff
@@ -0,0 +1,132 @@
+//! Reference-counted buffer data type.
+//!
+//! Requires some reference-counted type especially for frame buffer pools.
+//! `Arc` does not allow mutability with several references present,
```

Comment:
You can allow that with `Arc` in the same way as here by going via `unsafe` code. That said, `Arc` is more heavy as it also allows for weak references, so having a lightweight implementation might be worth it.

However I'm wondering how/why allowing mutable access to the content is in any way sound if there are multiple references.

Reply:
I agree with having a lightweight implementation for now, we can iterate over that later in my opinion.

About multiple references, it seems we cannot remove them because they are necessary for the `SimpleVideoFrame` struct

Reply:
Can you point me to the place in the code where that is needed? Maybe I have an idea :)

But never ever have public API that allows to get multiple mutable references to the same data at the same time. If anything this must be an unsafe function, and to get around UB it must return a raw `*mut` pointer and not a `&mut` reference. Having multiple mutable references to the same data at the same time is UB in Rust.

Reply:
I perfectly agree with you about your idea of APIs :)

You can find the problem at line 339 of the `frame.rs` file of this PR (`SimpleVideoFrame` has been added in this PR)

Reply:
I don't think this is the way to go there, that's unsound like this.

I see two options here:
* You use the normal `make_mut()` approach which copies the data if needed (>1 references). This should only copy if actually needed here as you borrow from the passed in frame anyway. So if the passed in frame has only a single reference then that stays like this.
* Go the GStreamer (and ffmpeg nowadays IIRC?) way and separate the actual data (in the end just some kind of byte array with the correct alignment etc, literally the same for audio, video, etc.) from the view on it (like `gst_video::VideoFrameRef` or `gst_audio::AudioFrameRef`, but see also the non-`Ref` variants which allow to take ownership instead of borrowing).
  The former is a ref-counted thing you can pass around, and you can "map" it to get access to the data as bytes: read-write (which creates a copy if someone else owns a reference, and otherwise not), and read-only (which allows multiple readers at once). This is basically like Rust's `Arc` without weak references overhead. 
  The latter would be helper API on top of all that that takes such a refcounted thing and "format information" to provide a convenient view into the underlying data (and read-only/read-write map the data).

You would in either case do copies of the data whenever needed (multiple writers) but the fundamental difference between the two is the in the first case even multiple readers need to copy, while in the second case only multiple writers (or readers+writers) would need to copy. In addition the second case seems slightly cleaner to me because you separate the "storage" (it's all bytes in the end) from how you interpret the data.

This would later also potentially allow having data not just stored in a `Vec<u8>` but e.g. in a `memfd` (which also has a concept of mapping btw) or whatever else while allowing access to it in a generic way.

Let me know if you want to discuss the second in a bit more detail.

Reply:
I like the second case because, as you said, data and its representation are two different things, thanks for the explanation! :)

Since @lu-zero proposed to integrate the `v_frame` crate, we could see if the second approach is already implemented in there and if not, trying to add it in some way